### PR TITLE
impl(pubsub): add start and end span for publisher tracing connection

### PR DESCRIPTION
**Pull request by alevenberg**
Part of #[12525](https://github.com/googleapis/google-cloud-cpp/issues/12525)

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12617)

### CODE REVIEW

File: google/cloud/pubsub/internal/publisher_tracing_connection.cc [hunk 1]
```diff
@@ -52,14 +115,19 @@ void PublisherTracingConnection::ResumePublish(ResumePublishParams p) {
...
+                                                      std::move(connection));
 }
 
 #else  // GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
 
 std::shared_ptr<pubsub::PublisherConnection> MakePublisherTracingConnection(
+    pubsub::Topic topic,
     std::shared_ptr<pubsub::PublisherConnection> connection) {
+  // Silence the unused parameter warnings.
+  (void)topic;
```
Comment (devbww) [inline google/cloud/pubsub/internal/publisher_tracing_connection.cc:130]: I believe you can just not give the parameter a name to achieve the same effect.

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc [hunk 2]
```diff
@@ -15,35 +15,134 @@
...
+}
+
+TEST(PublisherTracingConnectionTest, PublishSpanOnError) {
+  namespace sc = ::opentelemetry::trace::SemanticConventions;
+  auto span_catcher = InstallSpanCatcher();
+  auto mock = std::make_shared<MockPublisherConnection>();
+  EXPECT_CALL(*mock, Publish)
+      .WillOnce([&](pubsub::PublisherConnection::PublishParams const&) {
+        EXPECT_TRUE(ThereIsAnActiveSpan());
+        return google::cloud::make_ready_future(
```
Comment (coryan) [inline google/cloud/pubsub/internal/publisher_tracing_connection_test.cc:105]: We rarely fully qualify types in `google::cloud`.  Your call.

**Review by coryan — APPROVED** [review]
A few comments for you to consider, but I would be happy if the code got it as-is.

File: google/cloud/pubsub/internal/publisher_tracing_connection.cc [hunk 3]
```diff
@@ -13,15 +13,22 @@
 // limitations under the License.
 
 #include "google/cloud/pubsub/internal/publisher_tracing_connection.h"
+#include "google/cloud/pubsub/message.h"
 #include "google/cloud/pubsub/publisher_connection.h"
+#include "google/cloud/future_generic.h"
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection.cc:18]: Whoops. I think this was clang-tidy. Changed.

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc (hunk 2 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection_test.cc]: Changed.

File: google/cloud/pubsub/internal/publisher_tracing_connection.cc [hunk 4]
```diff
@@ -30,11 +37,67 @@ namespace pubsub_internal {
...
+  options.kind = opentelemetry::trace::SpanKind::kProducer;
+  auto span =
+      internal::GetTracer(internal::CurrentOptions())
+          ->StartSpan(
+              topic + " send",
+              {
+                  {sc::kMessagingSystem, "pubsub"},
+                  {sc::kMessagingDestinationName, topic},
+                  {sc::kMessagingDestinationTemplate, "topic"},
+                  {"messaging.pubsub.ordering_key", m.ordering_key()},
```
[the remark this answers is not in the document]
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection.cc:55]: That's fair. I think that means using a different OTel api though, so I will make those changes in a separate PR.

#12620

File: google/cloud/pubsub/internal/publisher_tracing_connection.cc [hunk 5]
```diff
@@ -30,11 +37,67 @@ namespace pubsub_internal {
...
+future<StatusOr<std::string>> EndPublishSpan(
+    opentelemetry::nostd::shared_ptr<opentelemetry::trace::Span> span,
+    future<StatusOr<std::string>> f) {
+  return f.then([span = std::move(span)](auto fut) {
+    auto message_id = fut.get();
+    Status const& status = message_id.status();
+    if (status.ok()) {
+      span->SetStatus(opentelemetry::trace::StatusCode::kOk);
+      span->SetAttribute("gcloud.status_code", 0);
+      span->SetAttribute("messaging.message_id", message_id.value());
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/publisher_tracing_connection.cc:71]: Consider just:

```cc
  return f.then([span = std::move(span)](auto fut) {
    auto message_id = fut.get();
    if (message_id) span->SetAttribute("messaging.message_id", *message_id);
    return EndSpan(span, message_id);
  }
```

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc [hunk 6]
```diff
@@ -15,35 +15,132 @@
...
-#include "google/cloud/internal/opentelemetry.h"
-#include "google/cloud/opentelemetry_options.h"
+#include "google/cloud/pubsub/topic.h"
+#include "google/cloud/future_generic.h"
+#include "google/cloud/internal/make_status.h"
+#include "google/cloud/status.h"
+#include "google/cloud/status_or.h"
 #include "google/cloud/testing_util/opentelemetry_matchers.h"
+#include "google/cloud/testing_util/status_matchers.h"
+#include "gtest/gtest.h"
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/publisher_tracing_connection_test.cc:28]: we should probably remove this include. Or use `<gtest/gtest.h>` if it is actually necessary.

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc [hunk 7]
```diff
@@ -15,35 +15,132 @@
...
+      Topic("test-project", "test-topic"), std::move(mock));
+
+  auto response = connection
+                      ->Publish({pubsub::MessageBuilder{}
+                                     .SetData("test-data-0")
+                                     .SetOrderingKey("ordering-key-0")
+                                     .Build()})
+                      .get();
+
+  EXPECT_THAT(response, StatusIs(StatusCode::kOk));
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/publisher_tracing_connection_test.cc:75]: nit: I think we tend to say `EXPECT_STATUS_OK(response)`

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc [hunk 8]
```diff
@@ -15,35 +15,132 @@
...
+              OTelAttribute<std::string>(sc::kMessagingSystem, "pubsub"),
+              OTelAttribute<std::string>(
+                  sc::kMessagingDestinationName,
+                  "projects/test-project/topics/test-topic"),
+              OTelAttribute<std::string>(sc::kMessagingDestinationTemplate,
+                                         "topic"),
+              OTelAttribute<std::string>("messaging.pubsub.ordering_key",
+                                         "ordering-key-0"),
+              OTelAttribute<int>("gcloud.status_code", 0),
+              OTelAttribute<size_t>("messaging.message.total_size_bytes", 45),
```
Comment (dbolduc) [inline google/cloud/pubsub/internal/publisher_tracing_connection_test.cc:92]: nit: I don't think `std::size_t` is a consistently sized attribute type: https://github.com/open-telemetry/opentelemetry-cpp/blob/cdbc90abf3b7170f898439fd34126469c42686c6/sdk/include/opentelemetry/sdk/common/attribute_utils.h#L30-L44

Maybe use `std::uint64_t`?

**Review by dbolduc — APPROVED** [review]
Just nits

File: google/cloud/pubsub/internal/publisher_tracing_connection.cc (hunk 5 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection.cc]: Nice. Much cleaner. Done.

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc (hunk 6 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection_test.cc]: Yeah.... I shouldn't accept clang-tidys help.

Whoops.

Removed.

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc (hunk 7 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection_test.cc]: Done

File: google/cloud/pubsub/internal/publisher_tracing_connection_test.cc (hunk 8 above)
Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection_test.cc]: It isn't. I tried `std::uint64_t` and it failed the m32 build. 

I'll try`std::uint32_t`.

Comment (alevenberg) [reply google/cloud/pubsub/internal/publisher_tracing_connection_test.cc]: From one of the failing builds
```
Step #3:   Actual: { [messaging.message_id=std::string:test-id-0, messaging.message.total_size_bytes=std::uint64_t:45 ]} 

Step #3: matcher #5: has a first field that is equal to "messaging.message.total_size_bytes", and has a second field that is a variant<> with value of type 'unsigned int' and the value is equal to 45
```
Gonna go back to `size_t`.

Let me know if there's a better way to do this. Maybe we could check the architecture size then set the type at run time?

**Pull request merged**